src/main/scala/com/snowplowanalytics/refererparser/Parser.scala

```diff
@@ -0,0 +1,166 @@
+ * methods to generate a Referer object from a
+ * referer URL.
+ */
+object Parser {
+  def create[F[_]: Sync](filePath: String): F[Either[Exception, Parser]] =
+    Sync[F].delay { Source.fromFile(filePath).mkString }.map(rawJson =>
+        ParseReferers.loadJsonFromString(rawJson).map(referers => new Parser(referers)))
+}
+
+/**
+ * Parser class - Scala version of Java Parser, with
+ * everything wrapped in Sync
+ */
+class Parser private (referers: Map[String, RefererLookup]) {
+
+  private def toUri(uri: String): Option[URI] = {
+    if (uri == "")
+      None
+    else
+      Either.catchNonFatal(new URI(uri)).toOption
+  }
+
+  def parse(refererUri: URI): Option[Referer] =
+    parse(refererUri, None, Nil)
+
+  def parse(refererUri: String): Option[Referer] =
+    toUri(refererUri).flatMap(uri => parse(uri, None, Nil))
+
+  def parse(refererUri: URI, pageHost: String): Option[Referer] =
+    parse(refererUri, Some(pageHost), Nil)
+
+  def parse(refererUri: String, pageHost: String): Option[Referer] =
+    toUri(refererUri).flatMap(uri => parse(uri, Some(pageHost), Nil))
+
+  def parse(refererUri: URI, pageUri: URI): Option[Referer] =
+    parse(refererUri, Some(pageUri.getHost), Nil)
+
+  def parse(refererUri: String, pageUri: URI): Option[Referer] =
+    toUri(refererUri).flatMap(uri => parse(uri, Some(pageUri.getHost), Nil))
+
+  /**
+   * Parses a `refererUri` URI to return
+   * either Some Referer, or None.
+   */
+  def parse(
+    refererUri: URI,
+    pageHost: Option[String],
+    internalDomains: List[String]
+  ): Option[Referer] = {
+    val scheme = refererUri.getScheme
+    val host = refererUri.getHost
+    val path = refererUri.getPath
+    val query = Option(refererUri.getQuery)
+
+    val validUri = (scheme == "http" || scheme == "https") && host != null && path != null
+
+    if (validUri) {
+      if ( // Check for internal domains
+        pageHost.map(_.equals(host)).getOrElse(false) ||
+        internalDomains.map(_.trim()).contains(host)
+      ) {
+        Some(InternalReferer)
+      } else {
+        Some(lookupReferer(host, path).map(lookup => {
+          lookup.medium match {
+            case UnknownMedium => UnknownReferer
+            case SearchMedium => SearchReferer(
+              lookup.source,
+              query.flatMap(q => extractSearchTerm(q, lookup.parameters)))
+            case InternalMedium => InternalReferer
+            case SocialMedium => SocialReferer(lookup.source)
+            case EmailMedium => EmailReferer(lookup.source)
+            case PaidMedium => PaidReferer(lookup.source)
+          }
+        }).getOrElse(UnknownReferer))
+      }
+    } else {
+      None
+    }
+  }
+
+  private def extractSearchTerm(query: String, possibleParameters: List[String]): Option[String] =
+    extractQueryParams(query).find(p => possibleParameters.contains(p._1)).map(_._2)
+
+  private def extractQueryParams(query: String): List[(String, String)] =
```

Comment:
I rewrote this to get rid of apache httpclient (issue #28) based on https://stackoverflow.com/questions/13592236/parse-a-uri-string-into-name-value-collection. Urls with anchors isn't a problem because it first goes through URI: `(new URI("http://snowplow.com/?name=alex#hello")).getQuery == "name=alex"`.

I could go back to the original but we'd need to add back the httpclient dependency.

Reply:
ah that was why, thanks for reminding me :+1:

Reply:
great work